### DISCUSSION TITLE
In [id].js a value of { id } is assigned to query. Why?

### DISCUSSION THREAD

**Discussion by hanzallahmsd** [vercel/next.js#34419] [Help]
Example with-mongodb-mongoose > In the [id].js in api\pets why the value of {id} is assigned to query if it is not used? Is it forgotten or I am not getting something?

**Comment by 2stash** [comment]
Good morning.  The [id].js is for a dynamic route. Meaning if you have '/blog/[id].js' as your pages structure, any url hits to '/blog/1', '/blog/something', will be routed to the file [id].js.

So using a hook like useRouter.

if you check the value of

`const queryParameters = useRouter()`

and 

`console.log(queryParameters.query)`

the value printed out is the path name. So '1' or 'something', as an object with the form {id: '1'}.

This is the same behavior if you have another file in your pages directory like '/blog/dashboard.js'. Any hits to '/blog/dashboard' will use the file dashboard.js. And assuming you only have [id].js, index.js, and dashboard.js, any paths that are not 'blog/dashboard' will use [id].js. "dashboard" will still be assigned to query.

I hope that helps.

**Comment by hanzallahmsd** [comment]
Well! Thanks! I was asking about line 5 in pages/api/pets/[id].js 
`const { query: { id }, method,} = req `

**Reply by 2stash** [reply]
@mmuhannad Good morning! Not exactly sure what you mean. But query: { id } is the dynamic portion of the url. You can you use it if you need it or not. In the mongoose example, the id is used to fetch data.

https://github.com/vercel/next.js/blob/canary/examples/with-mongodb-mongoose/pages/api/pets/%5Bid%5D.js

```
  switch (method) {
    case 'GET' /* Get a model by its ID */:
      try {
        const pet = await Pet.findById(id)
        if (!pet) {
          return res.status(400).json({ success: false })
        }
        res.status(200).json({ success: true, data: pet })
      } catch (error) {
        res.status(400).json({ success: false })
      }
      break
```

**Comment by pingustar** [comment]
I just noticed the same thing. 

`getServerSideProps` -> `context.resolvedUrl` has unexpected query params

I get for my page `/posts/[id].tsx` the following resolved url:

`/posts/123?id=123`

I would expect: 

`/posts/123`

Is there a way to disable this behaviour?

**Reply by 2stash** [reply]
@pingustar This is intended behavior, although there are plenty of discussion posts about it and improvement suggestions.

https://nextjs.org/docs/api-reference/data-fetching/get-server-side-props#context-parameter

"resolvedUrl: A normalized version of the request URL that strips the _next/data prefix for client transitions and includes original query values."

I didn't try this but one posts recommends using "req.url".

https://github.com/vercel/next.js/discussions/18885

**Reply by pingustar** [reply]
thanks @2stash 

unfortunately `req.url` does not  seem like a viable option, it looks like this: `/_next/data/development/posts/123.json?id=123`

**Reply by 2stash** [reply]
![Screen Shot 2022-02-17 at 8 34 26 AM](https://user-images.githubusercontent.com/60017468/154503490-75e88c3f-30bb-44f0-bc69-411ae5046751.png)

@pingustar 
Anyway you can you use the hook 'useRouter' directly or by calling it before and passing it as a prop?
```
  const router = useRouter();

```

asPath would be available.

In any case it seems like it's intended to be in that format. "/posts/123?id=123"
